docs/push/design.md

```diff
@@ -0,0 +1,497 @@
+handshake is not completed in the first try, it MUST disconnect the WebSocket
+and begin a new connection.
+
+**NOTE**: Applications may request registrations or unregistrations from the
+UserAgent, before or when the handshake is in progress. The UserAgent MAY buffer
+these or report errors to the application. But it MUST NOT send these requests
+to the PushServer until the handshake is completed.
+
+#### UserAgent -> PushServer
+
+<dl>
+    <dt>messageType = "hello"</dt>
+    <dd></dd>
+    <dt>uaid string REQUIRED</dt>
+    <dd>If the UserAgent has a previously assigned UAID, it should send it. Otherwise
+  send an empty string.</dd>
+    <dt>channelIDs list of strings REQUIRED</dt>
+    <dd>If the UserAgent has a list of channelIDs it wants to be notified of, it must
+  pass these, otherwise an empty list.</dd>
+    <dt>use_webpush bool OPTIONAL</dt>
+    <dd>If the UserAgent wants the WebPush data support for notifications it
+  may include this field with a value of *true*. This field may appear for legacy reasons.</dd>
+    <dt>broadcasts object OPTIONAL</dt>
+    <dd>If the UserAgent wants to subscribe to specific megaphone broadcasts, it may
+  include this field: an object consisting of broadcastIDs mapped to the
+  UserAgent's current broadcast versions.</dt>
+</dl>
+
+Extra fields: The UserAgent MAY pass any extra JSON data to the PushServer. This
+data may include information required to wake up the UserAgent out-of-band. The
+PushServer MAY ignore this data.
+
+**Example**
+
+```json
+{
+  "messageType": "hello",
+  "uaid": "fd52438f-1c49-41e0-a2e4-98e49833cc9c",
+  "broadcasts": {
+    "test/broadcast1": "v3",
+    "test/broadcast2": "v0"
+  }
+}
+```
+
+#### PushServer -> UserAgent
+
+<dl>
+    <dt>messageType = "hello"</dt>
+    <dd></dd>
+    <dt>uaid string REQUIRED</dt>
+    <dd>The UserAgent's provided UAID OR a new UAID.</dd>
+    <dt>status number REQUIRED</dt>
+    <dd>Used to indicate success/failure. MUST be one of:
+
+  * 200 - OK. Successful handshake.
+
+  * 503 - Service Unavailable. Database out of space or offline. Disk space
+    full or whatever other reason due to which the PushServer could not grant
+    this registration. UserAgent SHOULD avoid retrying immediately.</dd>
+    <dt>use_webpush bool OPTIONAL</dt>
+    <dd>Must match sent value</dd>
+    <dt>broadcasts object OPTIONAL</dt>
+    <dd>*Only* broadcasts the UserAgent has requested subscriptions to whose version
+values differ from the UserAgent's provided values. It may be empty if the
+UserAgent is up to date.</dd>
+</dl>
+
+**Example**
+
+```json
+{
+  "messageType": "hello",
+  "status": 200,
+  "uaid": "fd52438f-1c49-41e0-a2e4-98e49833cc9c",
+  "broadcasts": {
+    "test/broadcast2": "v1"
+  }
+}
+```
+
+
+### Register
+
+The Register message is used by the UserAgent to request that the PushServer
+notify it when a channel changes. Since channelIDs are associated with only one
```

Comment:
_editorial nit_: I wonder if we should change "associated" to "combined", since that's more accurate about what's going on.